homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,234 @@
+"""Module to coordinate llm tools."""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Callable, Iterable
+from dataclasses import dataclass
+from functools import cached_property
+import inspect
+import logging
+from types import FunctionType, NoneType, UnionType
+from typing import Any, get_args, get_type_hints
+
+import voluptuous as vol
+from voluptuous_openapi import UNSUPPORTED, convert
+
+from homeassistant.core import Context, HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+
+from . import config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_KEY = "llm_tool"
+
+
+@dataclass(slots=True)
+class ToolInput:
+    """Tool input to be processed."""
+
+    tool_name: str
+    tool_args: dict[str, Any]
+    platform: str
+    context: Context | None
+    user_prompt: str | None
+    language: str | None
+    agent_id: str | None
+    conversation_id: str | None
+    device_id: str | None
+    assistant: str | None
+
+
+class Tool:
+    """LLM Tool base class."""
+
+    name: str
+    description: str | None = None
+    parameters: vol.Schema = vol.Schema({})
+
+    @cached_property
+    def specification(self) -> dict[str, Any]:
+        """Get the tool specification in OpenAPI-compatible format."""
+        result = {"name": self.name}
+        if self.description:
+            result["description"] = self.description
+
+        def custom_serializer(schema: Any) -> Any:
+            """Wrap self.custom_serializer."""
+            return self.custom_serializer(schema)
+
+        result["parameters"] = convert(
+            self.parameters, custom_serializer=custom_serializer
+        )
```

Comment:
```suggestion
        result["parameters"] = convert(
            self.parameters, custom_serializer=self.custom_serializer
        )
```

Reply:
The `custom_serializer` should have one argument. As a class method, `self` is added automatically.

Reply:
I don't see why the wrapper function is needed.

Reply:
To avoid the Type error about missing required positional argument. Anyway, it is already removed.